### PR TITLE
Issue #8267: Implement full records support

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #8267: Implement full records support

This PR will add full support for the Java 14 records syntax.

From [JEP](https://openjdk.java.net/jeps/384):
_**Java Grammar**_
_RecordDeclaration:_
  _{ClassModifier} `record` TypeIdentifier [TypeParameters]_
    _RecordHeader [SuperInterfaces] RecordBody_

_RecordHeader:_
 _`(` [RecordComponentList] `)`_

_RecordComponentList:_
 _RecordComponent { `,` RecordComponent}_

_RecordComponent:_
 _{Annotation} UnannType Identifier_
 _VariableArityRecordComponent_

_VariableArityRecordComponent:_
 _{Annotation} UnannType {Annotation} `...` Identifier_

_RecordBody:_
  _`{` {RecordBodyDeclaration} `}`_

_RecordBodyDeclaration:_
  _ClassBodyDeclaration_
  _CompactConstructorDeclaration_

_CompactConstructorDeclaration:_
 _{Annotation} {ConstructorModifier} SimpleTypeName ConstructorBody_

Note that I've taken some liberties with the java grammar, in order to reuse existing code and reduce extraneous rules in the grammar file.

~~_Reviewers, please note that the commits from #8293 are temporarily on this PR, since I built from them. Once #8293 is merged, I will rebase._~~

### CODE REVIEW

**Comment by nmancus1** [comment]
This post will be used to collect check and AST regression reports.

**Comment by nmancus1** [comment]
This post will be used to collect possibly effected checks.

**Comment by pbludov** [comment]
@nmancus1 please fix ACL for the regression reports. Some of them are inaccessible.
```
$ curl -s https://checkstyle-reports-java14.s3.amazonaws.com/reports/full-records-support_check_diff_reports_2020_07_08/diff_part2/apache-struts/xref/home/nick/development/contribution/checkstyle-tester/repositories/apache-struts/plugins/plexus/src/main/java/org/apache/struts2/plexus/PlexusLifecycleListener.java.html#L33

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>23540C1B2840C8AB</RequestId><HostId>rdZhmgHVa5YNyVeNgzu742UFHMytj8DflZiJnFAm5s0IQnng3RCYZq6WGpYAX4ndkESzHLedKnM=</HostId></Error>
```

**Comment by esilkensen** [comment]
Hi @rnveach, friendly ping for this PR - if it looks good to you, would be great to get it merged so @nmancus1 can have a clean base to build from for the follow-up PR's addressing the affects checks.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 1]
```diff
@@ -3590,12 +3593,136 @@
...
+     * |       |--TYPE
+     * |       |   `--IDENT (Comp)
+     * |       `--IDENT (y)
+     * |--RPAREN ())
+     * `--OBJBLOCK
+     *      |--LCURLY ({)
+     *       `--RCURLY (})
+     * </pre>
+     *
+     * @since 8.36
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3644]: @romani how's this?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3644]: Good, please keep it.

**Comment by romani** [comment]
from JEP:
> A record can be declared top level ....

please create such input.
Or fix indentation at https://github.com/checkstyle/checkstyle/blob/386451993069f8e909e7611e457896503d8e6a56/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14Records.java#L72 (indentation fix is separate commit.)

> CompactConstructorDeclaration:
 {Annotation} {ConstructorModifier} SimpleTypeName ConstructorBody

Please put in Input different(all possible) modifiers for CompactConstructorDeclaration.
example of multiple c-tors that call each other.

**Comment by nmancus1** [comment]
> Please put in Input different(all possible) modifiers for CompactConstructorDeclaration.

From [JLS]( http://cr.openjdk.java.net/~gbierman/jep359/jep359-20191031/specs/records-jls.html#jls-8.10.5):

_The compact constructor must be declared public._

**Comment by romani** [comment]
please fix indentation of https://github.com/checkstyle/checkstyle/blob/386451993069f8e909e7611e457896503d8e6a56/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14Records.java#L71 

they should not be indented, in separate commit of this PR or separate PR

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 2]
```diff
@@ -3572,13 +3572,16 @@
     /**
      * The {@code record} keyword.  This element appears
      * as part of a record declaration.
+     *
+     * @since 8.35
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3576]: 8.35 was just released so this needs to be upped to 36

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3576]: This was added in 8.35 with the preliminary record support.  Should I make this a separate PR or commit?

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3576]: Sorry, didn't make the connection.

**Pull request merged**